pkg/storage/swift/utils.go

```diff
@@ -0,0 +1,94 @@
+package swift
+
+import (
+	"fmt"
+
+	"github.com/gophercloud/utils/openstack/clientconfig"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+
+	imageregistryv1 "github.com/openshift/cluster-image-registry-operator/pkg/apis/imageregistry/v1"
+	regopclient "github.com/openshift/cluster-image-registry-operator/pkg/client"
+	"github.com/openshift/cluster-image-registry-operator/pkg/clusterconfig"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+// setSwiftCredentials reads the credentials from the operator's native secret
+func setSwiftCredentials(listers *regopclient.Listers, cr *imageregistryv1.ImageRegistryConfigStorageSwift) error {
+	// Look for a user defined secret to get the Swift credentials
+	sec, err := listers.Secrets.Get(imageregistryv1.ImageRegistryPrivateConfigurationUser)
+	if err != nil {
+		return err
+	}
+
+	if v, ok := sec.Data["REGISTRY_STORAGE_SWIFT_USERNAME"]; ok {
+		cr.Username = string(v)
+	} else {
+		return fmt.Errorf("secret %q does not contain required key \"REGISTRY_STORAGE_SWIFT_USERNAME\"", fmt.Sprintf("%s/%s", imageregistryv1.ImageRegistryOperatorNamespace, imageregistryv1.ImageRegistryPrivateConfigurationUser))
+	}
+	if v, ok := sec.Data["REGISTRY_STORAGE_SWIFT_PASSWORD"]; ok {
+		cr.Password = string(v)
+	} else {
+		return fmt.Errorf("secret %q does not contain required key \"REGISTRY_STORAGE_SWIFT_PASSWORD\"", fmt.Sprintf("%s/%s", imageregistryv1.ImageRegistryOperatorNamespace, imageregistryv1.ImageRegistryPrivateConfigurationUser))
+	}
+
+	return nil
+}
+
+func parseClouds(data []byte) (map[string]clientconfig.Cloud, error) {
```

Comment:
you can do away with this function if you use the internal `gophercloud` utilities.

Reply:
I couldn't find a function there for this usecase. all of them try to find clouds.yaml file in the filesystem first and read data from there: findAndReadCloudsYAML()  https://github.com/gophercloud/utils/blob/master/openstack/clientconfig/utils.go#L99
But we need to parse data from the secret, and LoadCloudsYAML can't do that https://github.com/gophercloud/utils/blob/master/openstack/clientconfig/requests.go#L62

Reply:
See how I did it in the 2 links I posted in my previous comment. That should have an example of how you can do this.